scripts/main.js

```diff
@@ -0,0 +1,348 @@
+      }
+    }
+  }
+
+  return lose;
+}
+
+function newCell() {
+  const emptyCells = [];
+  for (let y = 0; y < 4; y++) {
+    for (let x = 0; x < 4; x++) {
+      if (!board[y][x]) {
+        emptyCells.push({ y: y, x: x });
+      }
+    }
+  }
+
+  const val = Math.floor(Math.random() * emptyCells.length);
+  board[emptyCells[val].y][emptyCells[val].x] = Math.random() < 0.9 ? 2 : 4;
+
+  if (checkIfLose()) {
+    message_lose.classList.remove("hidden");
+  }
+}
+
+function drawBoard() {
+  score.innerHTML = scoreVal;
+
+  for (let y = 0; y < 4; y++) {
+    for (let x = 0; x < 4; x++) {
+      if (board[y][x] === 0) {
+        rows[y].cells[x].innerHTML = "";
+      } else {
+        rows[y].cells[
+          x
+        ].innerHTML = `<div class="field-cell-num field-cell--${board[y][x]}">
+					${board[y][x]}
+					</div>`;
+      }
+    }
+  }
+}
+
+function moveUp() {
+  const originalBoard = board.map((row) => row.slice()); // Create a copy of the original board
+  let hasChanged = false; // Track if the board state has changed
+
+  for (let x = 0; x < 4; x++) {
+    let merged = [false, false, false, false]; // Array to track if a cell has already merged in this move
+
+    for (let y = 1; y < 4; y++) {
+      if (board[y][x] !== 0) {
+        // Check if the cell is not empty
+        let newY = y;
+        let moveDistance = 0; // Variable to track the distance the cell will move
+
+        // Move the cell up as far as possible
+        while (newY > 0 && board[newY - 1][x] === 0) {
+          newY--;
+          moveDistance++; // Increment move distance
+        }
+
+        // Check if we can merge the cell with the one above
+        if (
+          newY > 0 &&
+          board[newY - 1][x] === board[y][x] &&
+          !merged[newY - 1]
+        ) {
+          // Merge the cells by doubling the value
+          board[newY - 1][x] *= 2;
+          scoreVal += board[newY - 1][x]; // Increase the score
+          board[y][x] = 0; // Clear the original cell
+          merged[newY - 1] = true; // Mark this cell as merged
+          hasChanged = true;
+
+          // Animate the move with merging (both cells move to the same position)
+          const cellElement = rows[y].cells[x].firstChild;
+          if (cellElement) {
+            cellElement.classList.add(`move-up--${moveDistance + 1}`); // Add a class for animation
+          }
+        } else if (newY !== y) {
+          // If the cell moved but did not merge
+          board[newY][x] = board[y][x]; // Move the cell to its new position
+          board[y][x] = 0; // Clear the original position
+          hasChanged = true;
+
+          // Animate the move without merging
+          const cellElement = rows[y].cells[x].firstChild;
+          if (cellElement) {
+            cellElement.classList.add(`move-up--${moveDistance}`); // Add a class for the move animation
+          }
+        }
+      }
+    }
+  }
+
+  return hasChanged; // Return whether the board changed
+}
```

Comment:
The 'moveUp' function contains logic for merging cells and moving them up. Consider refactoring this logic into smaller functions to handle each task separately, which would improve code readability and maintainability.